Installation issue: llvm-amdgpu

### Steps to reproduce the issue

With the Dockerfile in https://gist.github.com/msimberg/2fba195244da1e5c38540d74b3a51c05:

```console
$ (podman|docker) build --build-arg BASE_IMAGE=docker.io/ubuntu:22.04 --build-arg SPACK_COMMIT=d0589285f72bc6ec9b5f804612448954dcdf819c -f Dockerfile .
```

### Error message

<details><summary>Error message</summary>

<pre>
==> Error: ProcessError: Command exited with status 1:
    'ninja' '-j3' '-v'

147 errors found in build log:
     12940    [833/904] Copying CXX header vector
     12941    [834/904] Copying CXX header version
     12942    [835/904] Copying CXX header wchar.h
     12943    [836/904] Copying CXX header wctype.h
     12944    [837/904] Building CXX object libcxxabi/src/CMakeFiles/cxxabi_static_objects.dir/cxa_virtual.cpp.o
     12945    [838/904] Building CXX object libcxxabi/src/CMakeFiles/cxxabi_static_objects.dir/cxa_aux_runtime.cpp.o
  >> 12946    FAILED: libcxxabi/src/CMakeFiles/cxxabi_static_objects.dir/cxa_aux_runtime.cpp.o
     12947    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/./bin/clang++ --target=x86_64-unknown-lin
              ux-gnu -DLIBCXX_BUILDING_LIBCXXABI -D_LIBCPP_BUILDING_LIBRARY -D_LIBCXXABI_BUILDING_LIBRARY -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__ST
              DC_LIMIT_MACROS -I/tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-src/libcxxabi/../libcxx/src -I/tmp/ro
              ot/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1 -I/tmp/root/spack-stage/spack-sta
              ge-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/x86_64-unknown-linux-gnu/c++/v1 -I/tmp/root/spack-stage/spack-st
              age-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-src/libcxxabi/include -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual
               -Wdelete-non-virtual-dtor -Wno-comment -fdiagnostics-color  -O3 -DNDEBUG -D_DEBUG -UNDEBUG -std=c++20 -MD -MT libcxxabi/src/CMakeFiles/cxxabi_sta
              tic_objects.dir/cxa_aux_runtime.cpp.o -MF libcxxabi/src/CMakeFiles/cxxabi_static_objects.dir/cxa_aux_runtime.cpp.o.d -o libcxxabi/src/CMakeFiles/c
              xxabi_static_objects.dir/cxa_aux_runtime.cpp.o -c /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-src/l
              ibcxxabi/src/cxa_aux_runtime.cpp
     12948    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-src/libcxxabi/src/cxa_aux_runtime
              .cpp:13:
     12949    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/new:
              93:
     12950    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/cstd
              lib:87:
     12951    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/stdl
              ib.h:93:
  >> 12952    /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/stdlib.h:38:12: error: no member named 'abort' in namespace 'std'
     12953    using std::abort;
     12954          ~~~~~^
  >> 12955    /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/stdlib.h:39:12: error: no member named 'atexit' in namespace 'std'
     12956    using std::atexit;
     12957          ~~~~~^
  >> 12958    /usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/stdlib.h:40:12: error: no member named 'exit' in namespace 'std'
     12959    using std::exit;
     12960          ~~~~~^
     12961    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-src/libcxxabi/src/cxa_aux_runtime
              .cpp:13:
     12962    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/new:
              93:
     12963    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/cstd
              lib:87:
  >> 12964    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/stdlib.h:148:34: error: un
              known type name 'ldiv_t'
     12965    inline _LIBCPP_INLINE_VISIBILITY ldiv_t div(long __x, long __y) _NOEXCEPT {
     12966                                     ^
  >> 12967    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/stdlib.h:149:12: error: no
               member named 'ldiv' in the global namespace
     12968      return ::ldiv(__x, __y);
     12969             ~~^
  >> 12970    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/stdlib.h:152:34: error: un
              known type name 'lldiv_t'
     12971    inline _LIBCPP_INLINE_VISIBILITY lldiv_t div(long long __x,
     12972                                     ^
  >> 12973    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/stdlib.h:154:12: error: no
               member named 'lldiv' in the global namespace
     12974      return ::lldiv(__x, __y);
     12975             ~~^
     12976    In file included from /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-src/libcxxabi/src/cxa_aux_runtime
              .cpp:13:
  >> 12977    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/new:346:11: error: no memb
              er named 'posix_memalign' in the global namespace
     12978      (void)::posix_memalign(&__result, __alignment, __size);
     12979            ~~^
  >> 12980    /tmp/root/spack-stage/spack-stage-llvm-amdgpu-5.5.1-dwwd3rnubq4ibpxpx6ptyo5v5ukrkgxy/spack-build-dwwd3rn/include/c++/v1/new:357:5: error: no type
              named 'free' in the global namespace
     12981      ::free(__ptr);
     12982      ~~^
     12983    9 errors generated.
...
</pre></details>


### Information on your system

* **Spack:** 0.23.0.dev0
* **Python:** 3.10.12
* **Platform:** linux-ubuntu22.04-icelake
* **Concretizer:** clingo

### Additional information

`spack-build-out.txt` and `spack-build-env.txt`: https://gist.github.com/msimberg/b722947c94a6598439d401a0d505bea4.

With the previous commit (https://github.com/spack/spack/commit/d079aaa08336d7805fc0361669e1d16b8d5de4bf) `llvm-amdgpu` builds fine (this commit seemed suspicous to me at first since it modifies patches for hipcc, but the patches apply for hipcc >= 6 and llvm-amdgpu is a dependency of hipcc so doesn't come into play here).

The first failing commit (https://github.com/spack/spack/commit/d0589285f72bc6ec9b5f804612448954dcdf819c) only seems to bump package versions and make minor formatting changes. I see nothing odd about the changes, so I can only assume it's one of the minor version bumps that cause a change. Does anything look suspicious to you or has any similar errors been seen elsewhere (I couldn't find anything relevant)? I can only assume one of the dependencies changes how llvm-amdgpu sets up its system include paths/definitions or something similar.

Comparing specs with the last working commit and the first broken commit I only see the following version changes:
- perl-module-build: 0.4232 -> 0.4234
- libbsd: 0.12.1 -> 0.12.2
- util-linux-uuid: 2.38.1 -> 2.40.1
- ninja: 1.11.1 -> 1.12.0

I haven't checked which (if any) of these version bumps on their own trigger the failure.

The error remains with the latest commit on develop that I tried today (https://github.com/spack/spack/commit/645c8eeaeb7ebc1d08686edb1bde465aafbcdb84).

The issues seems to affect at least llvm-amdgpu 5.5-5.7 and GCC 11 and 12 (the Dockerfile above uses GCC 11).

@srekolam @renjithravindrankannath @haampie

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
+++ b/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
@@ -102,6 +102,9 @@ class LlvmAmdgpu(CMakePackage, CompilerPackage):
 
     conflicts("^cmake@3.19.0")
 
+    # https://github.com/spack/spack/issues/45746
+    conflicts("^ninja@1.12:", when="@:5")
+
     root_cmakelists_dir = "llvm"
     install_targets = ["clang-tidy", "install"]
 

--- a/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
+++ b/var/spack/repos/builtin/packages/llvm-amdgpu/package.py
@@ -103,7 +103,7 @@ class LlvmAmdgpu(CMakePackage, CompilerPackage):
     conflicts("^cmake@3.19.0")
 
     # https://github.com/spack/spack/issues/45746
-    conflicts("^ninja@1.12:", when="@:5")
+    conflicts("^ninja@1.12:", when="@:6.0")
 
     root_cmakelists_dir = "llvm"
     install_targets = ["clang-tidy", "install"]

